guardian/typerighter
# Filter named entities from dictionary matches

## What does this change?

Right now, proper nouns cause false positives in Typerighter because they're not found in Collins Dictionary. This creates a lot of noise and erodes trust in the tool.

![Screenshot 2023-10-05 at 09 09 37](https://github.com/guardian/typerighter/assets/7423751/d5cc818d-c18a-4edf-81b7-45085aadf3f5)

This PR attempts to improve this by identifying named entities from text being checked and excluding them from dictionary matches. We use the named-entity recognition (NER) library [OpenNLP](https://opennlp.apache.org/) to achieve this. 

To start with, we use 3 models: people, organisations and locations. There is a full list of models here:

https://opennlp.sourceforge.net/models-1.5/

In the future, we might wish to use the model developed internally by the data science team as it was trained on Guardian content and will likely do a better job:

https://github.com/guardian/data-science-ner-service

We just went with OpenNLP as the simplest proof of concept.

## How to test

Run typerighter and composer locally, ensuring composer is pointing at the local typerighter instance as per [the docs](https://github.com/guardian/typerighter/blob/main/docs/03-composer-integration.md). Ensure the Collins Dictionary feature switch is enabled. Try adding some proper nouns and see if they are flagged as misspellings or not.

## How can we measure success?

Fewer false positives caused by proper nouns in text.

## Have we considered potential risks?

There is a risk that genuinely misspelled words might be incorrectly identified as named-entities and therefore excluded from spell-checking (i.e. false negatives).

## Images

TODO
